src/Cardano/CLI/Voting/Metadata.hs

```diff
@@ -181,53 +230,89 @@ metadataToJson = Api.metadataToJson Api.TxMetadataJsonNoSchema
     Just x   -> pure x
+  -- DECISION #10A:
+  --   deserialise the voting public key
   votePub   <- case Api.deserialiseFromRawBytes AsVotingKeyPublic votePubRaw of
-    Nothing -> throwError (_DeserialiseVotePublicKeyFailure # votePubRaw)
+    Nothing -> throwError (_MetadataParseFailure # (RegoVotingKey, "Failed to deserialise."))
     Just x  -> pure x
+  -- DECISION #10A:
+  --   deserialise the rewards address
   rewardsAddr <- case Api.deserialiseFromRawBytes Api.AsStakeAddress rewardsAddrRaw of
-    Nothing -> throwError (_DeserialiseRewardsAddressFailure # rewardsAddrRaw)
+    Nothing -> throwError (_MetadataParseFailure # (RegoRewardsAddress, "Failed to deserialise."))
     Just x  -> pure x
 
   let
     payload = mkVotePayload votePub stkVerify rewardsAddr slot
 
+  -- DECISION #11:
+  --   We found and deserialised the vote registration but the vote registration
+  --   signature is invalid.
   case payload `signVotePayload` sig of
-    Nothing   -> throwError (_MetadataSignatureInvalid # (payload, sig))
+    Nothing   -> throwError (_MetadataInvalidSignature # ())
     Just vote -> pure vote
 
   where
-    metaKey :: Word64 -> TxMetadata -> Either MetadataParsingError TxMetadataValue
-    metaKey key val@(TxMetadata map) =
-      case M.lookup key map of
-        Nothing     -> throwError (_MetadataMissingField # (val, key))
-        Just x      -> pure x
-
-    metaNum :: Integer -> TxMetadataValue -> Either MetadataParsingError TxMetadataValue
-    metaNum key val@(TxMetaMap xs) =
-      case find (\(k, v) -> k == TxMetaNumber key) xs of
-        Nothing     -> throwError (_MetadataValueMissingField # (val, key))
-        Just (_, x) -> pure x
-    metaNum key (x)            = throwError (_MetadataValueUnexpectedType # ("TxMetaMap", x))
-
-    asBytes :: TxMetadataValue -> Either MetadataParsingError ByteString
-    asBytes (TxMetaBytes bs) = pure bs
-    asBytes (x)              = throwError (_MetadataValueUnexpectedType # ("TxMetaBytes", x))
-
-    asInt :: TxMetadataValue -> Either MetadataParsingError Integer
-    asInt (TxMetaNumber int) = pure int
-    asInt (x)              = throwError (_MetadataValueUnexpectedType # ("TxMetaNumber", x))
+    metaValue
+        :: VoteRegistrationComponent
+        -> TxMetadata
+        -> Either MetadataParsingError TxMetadataValue
+    metaValue comp (TxMetadata map1) =
+        let
+            (k1, k2) = componentPath comp
+        in
+            case M.lookup k1 map1 of
+                Nothing ->
+                    throwError (_MetadataMissing # comp)
+                Just (TxMetaMap map2) ->
+                    case find (\(k, v) -> k == TxMetaNumber k2) map2 of
+                        Nothing ->
+                            throwError (_MetadataMissing # comp)
+                        Just (_, v) ->
+                            pure v
+                Just other ->
+                    throwError (_MetadataMissing # comp)
```

Comment:
```suggestion
    metaValue comp (TxMetadata map1) =
        case find (\(k, v) -> k == TxMetaNumber k2) <$> M.lookup k1 map1 of
            Just (TxMetaMap (Just (_, v))) -> pure v
            _ -> throwError (_MetadataMissing # componentPath comp)
```

Reply:
Didn't quite work out as simple as this, so I've just improved it somewhat in a56f54d.